src/Provider/AzureAD.php

```diff
@@ -0,0 +1,95 @@
+<?php
+/*!
+* Hybridauth
+* https://hybridauth.github.io | https://github.com/hybridauth/hybridauth
+*  (c) 2017 Hybridauth authors | https://hybridauth.github.io/license.html
+*/
+
+namespace Hybridauth\Provider;
+
+use Hybridauth\Adapter\OAuth2;
+use Hybridauth\Exception\UnexpectedApiResponseException;
+use Hybridauth\Data;
+use Hybridauth\User;
+
+/**
+ * AzureAD provider
+ */
+class AzureAD extends OAuth2
+{
+    /**
+    * {@inheritdoc}
+    */
+    public $scope = 'openid profile email offline_access https://graph.microsoft.com/User.Read';
+
+    /**
+    * {@inheritdoc}
+    */
+    protected function initialize()
+    {
+        parent::initialize();
+
+        $this->AuthorizeUrlParameters += [
+            'access_type' => 'offline'
+        ];
+
+        $this->tokenRefreshParameters += [
+            'client_id' => $this->clientId,
+            'client_secret' => $this->clientSecret
+        ];
+    }
+
+    protected function validateAccessTokenExchange($response)
+    {
+        $collection = parent::validateAccessTokenExchange($response);
+        if ($collection->exists('id_token')) {
+            $idToken = $collection->get('id_token');
+            //get payload from id_token
+            $parts = explode('.', $idToken);
+            list($headb64, $payload) = $parts;
+            // JWT token is base64url encoded
+            $data = base64_decode(str_pad(strtr($payload, '-_', '+/'), strlen($payload) % 4, '=', STR_PAD_RIGHT));
```

Comment:
HA v3 has "firebase/php-jwt" as dependency, does it make sense to use it here instead of custom solution?